implementation of Operations for devices

DarkPitoune opened the pull request:
What I consider to be the boilerplate that I will copy/paste for the rest of the file, so any comment is welcome.
Since I'm hella confused with traits/struct/impl I'm curious about any comment (see dubious name choices).
Also, ownership is handled by copy pasting the rust helper so it may inaccurate.

Leonils committed:
Retrieve config and propagate error if any

Leonils committed:
Add device list to global config

Leonils committed:
Load devices from config toml

Leonils committed:
Build device from toml table to include name

Leonils committed:
Verify that multiple device are well parsed

Leonils committed:
Verify device factory not found case and refactor to remove unwrap

Leonils committed:
Collect all errors when parsing the config

Leonils committed:
Pass name as a mandatory argument of build_from_toml_table

Leonils committed:
Test a second type of device with parameter

Leonils committed:
Implement build_from_toml_table for mountedFolder

Leonils committed:
Raise error if there are 2 devices with same name

Leonils committed:
Refactor global config loading

Leonils committed:
Implement add device to config

Leonils committed:
Add write config file method to GlobalConfigProvider

Leonils committed:
Use automock to mock the global config provider

Leonils committed:
Add a save method on global config

Leonils committed:
Implement save devices in config

Leonils committed:
Verify multiple device config saving

Leonils committed:
Implement serialization of mounted folder to toml

Leonils committed:
Add the global_config_provider to device operations

Leonils committed:
Implement listing devices from non empty config

Leonils committed:
Implement adding a device to empty config

Leonils committed:
Test when adding a device to config with another device already

Leonils committed:
Implement remove device in global config

Leonils committed:
Implement operation remove device by name

DarkPitoune committed:
first attempt, draft to discuss code formalism

Leonils committed:
remove _dir at the end of global config provider trait

Leonils committed:
Fix unused imports

DarkPitoune committed:
encapsulate in a struct and take in arguments

DarkPitoune committed:
Merge branch 'main' into pitoune/cli

DarkPitoune committed:
new features, get version, get help and invalid arguments

DarkPitoune commented:
![image](https://github.com/Leonils/hibernacli/assets/86973986/233a5f4b-6a1b-45c0-890c-c5210e65a12f)


Still kind of confused with the workflow and how to mock this but I feel like I can actually get it working once we agree on it..

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,233 @@
+        println!("{}", message);
+        Ok(())
+    }
+    fn read(&mut self) -> Result<String, String> {
```
you don't mutate  self here, so you can remove mut (and many other places that use mut because they use write and read)

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,233 @@
+
+struct Console;
+
+#[cfg_attr(test, automock)]
```
Here it would be more logical to mock UserInterface instead. You would have a MockUserInterface that is not linked to the console at all.

Leonils committed:
Replace &Box<dyn Device> by Rc<dyn device> to be able to mock DeviceOperations

DarkPitoune committed:
Merge pull request #12 from Leonils/debug/add-automock-to-operations

Replace &Box<dyn Device> by Rc<dyn device> to be able to mock DeviceOperations

Leonils reviewed (commented):
I think you don't give you enough credit Pierre. 

This is quite good. You defined the trait of your dependency, build your struct with some data (the dependency) and some behaviour, and tested behavior mocking your dependencies. That's about it. Good job. Continue on this track. 

Juste a little too much mut (but I know it came from what I told you when you started writing the mock manually and wanted to save the calls). You just don't need that anymore. 

And automock is more appropriate on traits.

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,233 @@
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+trait UserInterface {
+    fn write(&mut self, message: String) -> Result<(), String>;
```
You may also want to accept a &str instead of a String here. That would avoid a lot of conversions. You always can pass from String to &str by referencing the string (as long as you can borrow it)

But passing from a &str to a String requires taking ownership over the &str, I guess by some form of copy. If you don't need ownership or mutability, &str is best.

DarkPitoune committed:
initialize all the objects in run function
i made several things public, is that good ?
my file is starting to get pretty big, is that ok ?

DarkPitoune committed:
change types to have cleaner code
implement what leo nils told me :)

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,354 @@
+        }
+    }
+
+    fn display_message(&mut self, message: &String) {
```
So here you can accept a &str

Leonils replied:
And you don't need to have self mutable anymore

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,354 @@
+        }
+    }
+
+    fn display_help(&mut self) {
+        let _ = self.display_message(&HELP.to_string());
```
Therefore here, you don't need a mutable reference, and you can pass directly `display_message(HELP)`

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,354 @@
+        }
+
+        match args[2].as_str() {
+            "ls" | "list" => self.display_device_list(),
```
nice

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,354 @@
+        DEFAULT_CONFIG,
+    );
+    let operations = Operations::new(Box::new(local_file_storage));
+    let mut command_runner = CommandRunner {
```
You will probably need to call `register_device_factory` on concrete Operations struct. It is the part that is doing the injection of the factories with all concrete types.

DarkPitoune committed:
register the device ?? I'm not sure here

DarkPitoune committed:
device creation testing

DarkPitoune committed:
tests for all device operations successes

DarkPitoune commented:
At that point :
for all device operations, a test is developed. I haven't code much code for failing operations. I feel like there are many path to tests here , each one with a different mock.. Is there a way to do this in a clean way ?
On that same subject, I tried to create small functions with unique responsabilities but still my code feels heavy and not so readable. I'm looking for advice reagrding this. Some errors are not caught.. I guess this is to be done.
I'd like to merge this before moving on to the project management part to do it in a different PR.

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,562 @@
+
+    fn ask_for_string(&self, message: &str) -> String {
+        self.display_message(message);
+        match self.read_string() {
+            Ok(answer) => answer,
+            Err(_) => self.ask_for_string(message),
+        }
```
This pattern is usually avoidable with the map_err function: 

```
self.read_string().map_err(|| self.ask_for_string(message))
```

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,562 @@
+        }
+    }
+
+    fn create_new_device(&mut self, key: &DeviceFactoryKey) {
```
Maybe you can make this function return a Result, so you have any error, you can use the "?" operator to raise it up, and at the location where you run the command, you can handle almost all errors the same way, for instance printing them in red. 

Typically you would have: 

```rs
fn create_new_device(...) -> Result<(), String> {
   ...
   let factory = self....get_device_factory(...).ok_or("No factory found with key ...")
   while(...) { ... }
   let device = factory.build()?;
   self.operations.add_device(device)?
   self.display_message("OK")
}
```

And way up, when you run command, you do the matching: 

```rs
let command_result = run_device_command(args).or_else(|e| self.display_error(e))
```

Leonils reviewed (commented):
A first few comments to help you with the simplification of your code. I will go deeper soon, especially on simplification of your tests, but you can start working on these

DarkPitoune committed:
better error handling

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,558 @@
+Commands:
+    help              Display this help message
+    --version or -v   Display the version of the application
+    device [opt]      Manage devices
```
I think you may want to list commands here

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,558 @@
+
+    fn ask_question(&self, question_type: &QuestionType, question_statement: &str) -> String {
+        match question_type {
+            QuestionType::String => self.ask_for_string(question_statement),
```
Warning: the only device we have for now also have a UnixPath question: 

```rs
impl MountedFolderFactory {
    pub fn new() -> MountedFolderFactory {
        MountedFolderFactory {
            path_question: Question::new(
                "What is the path of the folder?".to_string(),
                QuestionType::UnixPath,
            ),
            name_question: Question::new(
                "How would you name this device?".to_string(),
                QuestionType::String,
            ),
            step: 0,
        }
    }
```

DarkPitoune replied:
what level of verification should i do ? regex or actually check the path exists or nothing ?

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,558 @@
+    fn display_device_list(&mut self) {
+        self.display_message("Device list:");
+        let devices = self.operations.list();
+        match devices {
+            Ok(devices) => {
+                for device in devices {
+                    self.display_message(&format!("Device: {}", device.get_name()));
+                }
+            }
+            Err(e) => {
+                self.display_message(&e);
+            }
+        }
```
here also I think you can improve error handling

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,558 @@
+            .get_available_device_factories()
+            .iter()
+            .find(|&key| key.key == device_key)
+        {
+            Some(key) => self.create_new_device(key),
+            None => {
+                self.display_message("No such device configuration exists");
+                return Ok(());
+            }
+        }
```
this too can be simplified by extending the pipeline

Leonils commented on src/cli.rs:
```diff
@@ -1,3 +1,558 @@
+            return;
+        }
+        let device_name = args[3].as_str();
+        match self.operations.remove_by_name(device_name.to_string()) {
+            Ok(_) => {
+                self.display_message("Removed device successfully");
+            }
```
This too. Here you can even map the ok to the succes message, map the error to the error message: 

```rs
let message = self
            .operations
            .remove_by_name(device_name.to_string())
            .map(|_| "Removed device successfully")
            .unwrap_or("Failed to remove device");

        self.display_message(message);
```

Leonils reviewed (commented):
A few more comments, but we are close.

DarkPitoune committed:
improve error handling

Leonils reviewed (approved):
Looks good for now. We will push improvements later. In my opinion when you are ready you can merge you PR.
